harness_tests/diagnostic_test/main1.c

```diff
@@ -1,205 +1,478 @@
-    double bt_on = read_current(MEAS_BT_IOUT);
-    double battery_on = read_bat_current(MEAS_PACK_IOUT);
-    ASSERT_FP_GREATER(bt_on, bt_init);
-    ASSERT_FP_LESS(battery_on, battery_init);
+    set_raw_heater_setpoint(&heater_1_shadow_setpoint, 0xFFF);
+    set_raw_heater_setpoint(&heater_1_sun_setpoint, 0xFFF);
+
+    // Check current due to both heaters on
+    _delay_ms(1000);
+    heater_on_curr_pack = measure_heater_current(MEAS_PACK_IOUT);
+    heater_on_curr_boost = measure_heater_current(MEAS_BT_IOUT);
 
+    /* Assert correct ranges */
+    ASSERT_FP_GREATER(heater_on_curr_pack, heater_off_curr_pack + 0.3);
+    ASSERT_FP_LESS(heater_on_curr_pack, heater_off_curr_pack + 0.4);
 
-    // Turn heaters off
+    ASSERT_FP_GREATER(heater_on_curr_boost, heater_off_curr_boost + 0.24);
+    ASSERT_FP_LESS(heater_on_curr_boost, heater_off_curr_boost + 0.32);
+
+    /* Turn heaters 1 and 2 off */
     set_raw_heater_setpoint(&heater_1_shadow_setpoint, 0);
     set_raw_heater_setpoint(&heater_1_sun_setpoint, 0);
     set_raw_heater_setpoint(&heater_2_shadow_setpoint, 0);
     set_raw_heater_setpoint(&heater_2_sun_setpoint, 0);
-
-    _delay_ms(1000);
-    
-    // Check current
-    double bt_off = read_current(MEAS_BT_IOUT);
-    double battery_off = read_bat_current(MEAS_PACK_IOUT);
-    // Should be approximately equal to when the heaters were off
-    ASSERT_FP_LESS(bt_off, bt_on);
-    ASSERT_FP_GREATER(battery_off, battery_on);
 }
 
+/* Tests calibrated and uncalibrated gyroscope values */
 void imu_test(void) {
-    // TODO: take measurements multiple times, make sure at least one of them is not 0
-    uint16_t raw_cal_x = 0, raw_cal_y = 0, raw_cal_z = 0;
-    ASSERT_TRUE(get_imu_cal_gyro(&raw_cal_x, &raw_cal_y, &raw_cal_z));
-    double cal_x = imu_raw_data_to_gyro(raw_cal_x);
-    double cal_y = imu_raw_data_to_gyro(raw_cal_y);
-    double cal_z = imu_raw_data_to_gyro(raw_cal_z);
-
-    ASSERT_FP_GREATER(cal_x, -0.1);
-    ASSERT_FP_LESS(cal_x, 0.1);
-    ASSERT_FP_NEQ(cal_x, 0);
-    ASSERT_FP_GREATER(cal_y, -0.1);
-    ASSERT_FP_LESS(cal_y, 0.1);
-    ASSERT_FP_NEQ(cal_y, 0);
-    ASSERT_FP_GREATER(cal_z, -0.1);
-    ASSERT_FP_LESS(cal_z, 0.1);
-    ASSERT_FP_NEQ(cal_z, 0);
-
-
-    uint16_t raw_uncal_x = 0, raw_uncal_y = 0, raw_uncal_z = 0;
-    ASSERT_TRUE(get_imu_uncal_gyro(&raw_uncal_x, &raw_uncal_y, &raw_uncal_z, NULL, NULL, NULL));
-    double uncal_x = imu_raw_data_to_gyro(raw_uncal_x);
-    double uncal_y = imu_raw_data_to_gyro(raw_uncal_y);
-    double uncal_z = imu_raw_data_to_gyro(raw_uncal_z);
-    ASSERT_FP_GREATER(uncal_x, -0.1);
-    ASSERT_FP_LESS(uncal_x, 0.1);
-    ASSERT_FP_NEQ(uncal_x, 0);
-
-    ASSERT_FP_GREATER(uncal_y, -0.1);
-    ASSERT_FP_LESS(uncal_y, 0.1);
-    ASSERT_FP_NEQ(uncal_y, 0);
-    
-    ASSERT_FP_GREATER(uncal_z, -0.1);
-    ASSERT_FP_LESS(uncal_z, 0.1);
-    ASSERT_FP_NEQ(uncal_z, 0);
+    uint16_t raw_data_imu = 0;
+    double gyr_data = 0;
+    uint8_t not_zero_flag = 0;
+
```

Comment:
You could probably refactor these 6 blocks below into a common function and call it 6 times with a different field number passed as an argument each time.

Reply:
Decided against this, as it would be easier to trace debug statements for this section when separate since there are only two non-assertion lines per block that could be easily condensed.